ystia/yorc
# Feature/gh 461 openstack boot volume

# Pull Request description

## Description of the change

### What I did

Allow to create an OpenStack Compute Instance using a bootable Volume, while previously the only possibility was to rely on the implicit use of a  use local disk on the hypervisor.

### How I did it

Added an optional property `boot_volume` to OpenStack compute instances.

Details of changes file by file :

`commands/bootstrap/inputs.go`:
Reduced complexity of functions, added a function getDatatypePropertyValue() to be able to interactively ask for inputs of a Datatype property.

`commands/bootstrap/inputs_test.go`:
New test testing getPropertiesInput() asking interactively values for string/boolean/list/datatype properties. A test function expectOutputsSendInputs() takes care of waiting for a question and providing the input value.

`commands/bootstrap/topology.go`:
Fixed an issue where Alien4Cloud REST API fails to manage boolean values in on-demand resources properties, passing strings instead.
Fixed an issue of yaml marshaling of map keys imageName and flavorName which are converted to lower case, which yaml files expect non lower case properties.

`commands/bootstrap/topology_test.go`:
Added tests createint a topology from templates

`commands/bootstrap/tosca_resources.go`:
Functions used to find tosca resources in bundled zip. These functions are not used in unit tests thanks to a build tag !testing
 
`commands/bootstrap/tosca_resources_testing.go`:
Functions defined only when running unit tests thanks to a build tag testing, to find TOSCA definitions locally instead of in the bundled zip

`data/tosca/yorc-openstack-types.yml`:
Added a datatype yorc.datatypes.openstack.BootVolume and a compute instance optional property boot_volume

`deployments/nodes.go`:
Added a function returns the string value of a property with optional nested keys

`prov/terraform/openstack/bs_volume.go`:
Compute a terraform boot volume resource if defined

`prov/terraform/openstack/bs_volume_test.go`:
New test testing the new function computeBootVolume()

`prov/terraform/openstack/consul_test.go`:
Calling new tests

`prov/terraform/openstack/osinstance.go`:
Fixed previously existing coding pratice issues reported by Sonar
Added code to create a boot volume if requested

`prov/terraform/openstack/osinstance_test.go`:
Added tests creating an instance with a boot volume
 
`prov/terraform/openstack/resources.go`:
Fixed previously existing coding pratice issues reported by Sonar
Added a new resource type BootVolume 

`prov/terraform/openstack/testdata/BootVolumeWrongSize.yaml`:
New test file providing a template with an incorrect value for the boot volume size of a compute instance

`prov/terraform/openstack/testdata/BootVolumeWrongType.yaml`:
New test file providing a template with an incorrect type for the value of boot volume property

`prov/terraform/openstack/testdata/OSInstanceWithBootVolume.yaml`:
New test file providing a template with Compute instance needing a boot volume creation.


### How to verify it

Download and unzip/untar yorc bundle at https://ystia.jfrog.io/ystia/binaries/ystia/yorc/dist/PR-483/yorc-4.0.0-SNAPSHOT.tgz
Use a bootstrap values file similar as usual, except from the Compute instance inputs, where you can provide these settings that will specify you want to create a bootable volume of 20 GB from a CentOS 7 image whose uuid is provided below:
```yaml
compute:
  flavorName: m1.medium
  key_pair: yorc
  boot_volume:
    uuid: a460db41-e574-416f-9634-96f2862f10fe
    source: image
    destination: volume
    size: 20 GB
    delete_on_termination: true
```
And use the following yorc and plugin download URLs:
```yaml
yorc:
  download_url: https://ystia.jfrog.io/ystia/binaries/ystia/yorc/dist/PR-483/yorc-4.0.0-SNAPSHOT.tgz

yorc_plugin:
  download_url: https://ystia.jfrog.io/ystia/binaries/ystia/yorc-a4c-plugin/dist/PR-143/alien4cloud-yorc-plugin-4.0.0-SNAPSHOT.zip
```
Bootstrap Yorc on OpenStack using this values file:

```
$ ./yorc bootstrap -v myvalues.yaml -n test
```
Once the boostrap has been succesfuly deployed,
login on the OpenStack Dashboard Console UI, select `Compute` > `Volume` and check a volume is attached to your newly created compute instance.

Login on the newly created compute instance and check the root disk is a disk of 20 GB :

```
$ lsblk
NAME   MAJ:MIN RM SIZE RO TYPE MOUNTPOINT
vda    253:0    0  20G  0 disk
└─vda1 253:1    0  20G  0 part /
```

Undeploy the bootstrapped environment :
```
$ ./yorc d undeploy -p test
```

Once done, login again on the OpenStack Dashboard Console UI, select  `Compute` > `Volume` and check the volume was deleted.

### Description for the changelog

Should support the creation of OpenStack Compute instances using bootable volume ([GH-461](https://github.com/ystia/yorc/issues/461))

## Applicable Issues

Fixes #461
